### PR TITLE
*: add proposalsCommitted metrics

### PR DESCRIPTION
**Pull request by xiang90**
/cc @fabxc @heyitsanthony @joshix @brian-brazil

### CODE REVIEW

File: etcdserver/raft.go [hunk 1]
```diff
@@ -226,6 +226,10 @@ func (r *raftNode) start(s *EtcdServer) {
 				if err := r.storage.Save(rd.HardState, rd.Entries); err != nil {
 					plog.Fatalf("raft save state and entries error: %v", err)
 				}
+				if !raft.IsEmptyHardState(rd.HardState) {
+					proposalsCommitted.Set(float64(rd.HardState.Commit))
```
Comment (xiang90) [inline etcdserver/raft.go]: @fabxc I do not think this is good as suggested in the godoc of prom's go-client... But on the other hand, it is very inconvenient to calculate the diff...

File: Documentation/metrics.md [hunk 2]
```diff
@@ -24,6 +24,7 @@ All these metrics are prefixed with `etcd_server_`
 |---------------------------|----------------------------------------------------------|---------|
 | has_leader                | Whether or not a leader exists. 1 is existence, 0 is not.| Gauge   |
 | leader_changes_seen_total | The number of leader changes seen.                       | Counter |
+| proposals_committed_total | The total number of consensus proposals get committed.   | Counter |
```
Comment (brian-brazil) [inline Documentation/metrics.md]: s/get//

File: Documentation/metrics.md [hunk 3]
```diff
@@ -32,6 +33,8 @@ is totally unavailable.
 
 `leader_changes_seen_total` counts the number of leader changes the member has seen since its start. Rapid leadership changes impact the performance of etcd significantly. It also signals that the leader is unstable, perhaps due to network connectivity issues or excessive load hitting the etcd cluster.
 
+`proposals_committed_total` counts the total number of consensus proposals get committed. This counter should be seen to be increasing over time if a cluster is healthy. Healthy members of an etcd cluster can still have different number of committed proposals at the same time. A member may be recovering from peers after startup, or lagging behind the leader in the cluster, or it may be up-to-date and participating fully, and it may be the leader. It is important to monitor this metrics across all the members in the cluster, a constant large lagging between a member and its leader indicates the member is slow or unhealthy.
```
Comment (brian-brazil) [inline Documentation/metrics.md]: s/lagging/lag/

File: etcdserver/raft.go (hunk 1 above)
Comment (fabxc) [inline etcdserver/raft.go]: This is fine as long as the value satisfies the counter semantics. In the end, if the diff was negative one couldn't do much about it either.
It's what one basically has to do for all exporters as well.

Comment (brian-brazil) [inline etcdserver/raft.go]: If you find yourself doing this, it hints this metric isn't being updated in the right place.

In this case as you're talking about the absolute values being useful, this is probably actually a gauge. For a counter you only really care about it's rate of increase.

File: Documentation/metrics.md (hunk 3 above)
Comment (fabxc) [inline Documentation/metrics.md]: s/get commited/that got commited/ or just leave the get?

File: etcdserver/raft.go (hunk 1 above)
Comment (xiang90) [inline etcdserver/raft.go]: ack. thanks!

File: Documentation/metrics.md (hunk 2 above)
Comment (fabxc) [inline Documentation/metrics.md]: As below

File: etcdserver/raft.go (hunk 1 above)
Comment (xiang90) [inline etcdserver/raft.go]: @brian-brazil Here is the problem, for leader we can know the increment. for followers, this value can be bumped (due to receive a snapshot) suddenly. it is just inconvenient to modify this metrics at multiple places.

Comment (brian-brazil) [inline etcdserver/raft.go]: That sounds very much like a gauge then, as it doesn't start at 0 per process.

It's the same way we'd consider the size of a append-only file to be a gauge.

Comment (xiang90) [inline etcdserver/raft.go]: > as it doesn't start at 0 per process.

No. it does not.

ok. will change this to gauge.

File: Documentation/metrics.md [hunk 4]
```diff
@@ -32,6 +33,8 @@ is totally unavailable.
 
 `leader_changes_seen_total` counts the number of leader changes the member has seen since its start. Rapid leadership changes impact the performance of etcd significantly. It also signals that the leader is unstable, perhaps due to network connectivity issues or excessive load hitting the etcd cluster.
 
+`proposals_committed_total` counts the total number of consensus proposals committed. This counter should be seen to be increasing over time if a cluster is healthy. Healthy members of an etcd cluster can still have different number of committed proposals at the same time. A member may be recovering from peers after startup, or lagging behind the leader in the cluster, or it may be up-to-date and participating fully, and it may be the leader. It is important to monitor this metrics across all the members in the cluster, a constant large lag between a member and its leader indicates the member is slow or unhealthy.
```
Comment (joshix) [inline Documentation/metrics.md]: s/this metrics/this metric/
s/cluster,/cluster./
s/a constant/A consistently/

**Comment by xiang90** [comment]
All fixed. PTAL

File: Documentation/metrics.md [hunk 5]
```diff
@@ -32,6 +33,8 @@ is totally unavailable.
 
 `leader_changes_seen_total` counts the number of leader changes the member has seen since its start. Rapid leadership changes impact the performance of etcd significantly. It also signals that the leader is unstable, perhaps due to network connectivity issues or excessive load hitting the etcd cluster.
 
+`proposals_committed_total` counts the total number of consensus proposals committed. This counter should be seen to be increasing over time if a cluster is healthy. Healthy members of an etcd cluster can still have different number of committed proposals at the same time. A member may be recovering from peers after startup, or lagging behind the leader in the cluster, or it may be up-to-date and participating fully, and it may be the leader. It is important to monitor this metric across all the members in the cluster. A consistently large lag between a member and its leader indicates the member is slow or unhealthy.
```
Comment (heyitsanthony) [inline Documentation/metrics.md]: This counter should increase over time if the cluster is healthy. Several healthy members of an etcd cluster may have a different total committed proposals at once. This discrepancy may be due to recovering from peers after starting, lagging behind the leader, or being the leader and therefore having the most commits. It is important to monitor this metric across all the members in the cluster; a consistently large lag between a single member and its leader indicates that member is slow or unhealthy.

**Comment by xiang90** [comment]
@heyitsanthony All fixed.

File: Documentation/metrics.md (hunk 5 above)
Comment (heyitsanthony) [inline Documentation/metrics.md]: oops--
s/may have a different total committed proposals/may have different total committed proposals/

**Comment by heyitsanthony** [comment]
lgtm after fixing last nit

**Comment by joshix** [comment]
lgtm

**Pull request merged**